client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java

```diff
@@ -168,6 +193,11 @@ public void stop() {
             if (stop.getAndSet(true)) {
                 return;
             }
+            if (!isReadingFromSocket.get() && !isCallbackExecuting.get()) {
+                // wait until we have completed the current call to the reply processors.
+                Exceptions.handleInterrupted(hasStopped::await);
```

Comment:
If this were moved to after the `closeQuietly` call below, couldn't we avoid the need for these two atomic bools?

Reply:
Thanks, I have simplified the code.

Reply:
Thanks. I have updated the fix . The code also needs to handle scenarios where stop() is invoked by `ClientSocketReaders` thread.